### PR TITLE
New Relic object management through context

### PR DESCRIPTION
**Pull request by izaurio**
* Manage transaction through context
* Wrap http.Handler with transaction

### CODE REVIEW

File: newrelicutil.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package newrelicutil
+
+import (
+	"context"
```
Comment (mkabischev) [inline newrelicutil.go:4]: please sort imports

File: newrelicutil.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/newrelic/go-agent"
+	"net/http"
+)
+
+type nrkey int
```
Comment (mkabischev) [inline newrelicutil.go]: call it just `key` or `contextKey`

File: newrelicutil.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+// WithTransaction puts the New Relic Transaction object to the given context
+// and returns the new context.
+func WithTransaction(ctx context.Context, txn newrelic.Transaction) context.Context {
+	return context.WithValue(ctx, transaction, txn)
+}
+
+// WrapHandlerCtx return the given http handler that is wrapped to New Relic Transaction.
+// Current New Relic Transaction is placed in the context.
+func WrapHandlerCtx(app newrelic.Application, name string, handler http.Handler) http.Handler {
```
Comment (mkabischev) [inline newrelicutil.go]: Maybe it's better to call it `Middleware`? And usage will be:
```
h := newrelic.Middleware("applicationName", "name", anotherHandler)
```
I think it's little more verbose.

File: newrelicutil.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		txn := app.StartTransaction(name, w, r)
+		defer txn.End()
+
+		ctx := r.Context()
+		ctx = WithTransaction(ctx, txn)
```
Comment (mkabischev) [inline newrelicutil.go]: You can combine two lines in single:
```
ctx := WithTransaction(r.Context(), txn)
```

File: newrelicutil_test.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package newrelicutil_test
+
+import (
+	"context"
```
Comment (mkabischev) [inline newrelicutil_test.go:4]: please sort imports

File: example_test.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package newrelicutil_test
+
+import (
+	"github.com/gettaxi/newrelicutil"
+	"net/http"
+)
+
+func ExampleWrapHandlerCtx() {
+	nrapp := newApp()
```
Comment (mkabischev) [inline example_test.go]: Let's move it to `WrapHandlerCtx` doc block. It will be visible in godoc.

File: newrelicutil_test.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+func TestTransactionCtx(t *testing.T) {
+	nrapp := newApp()
+	txn := nrapp.StartTransaction("foo", nil, nil)
+	ctx := context.Background()
+
+	if want, have := newrelic.Transaction(nil), newrelicutil.Transaction(ctx); want != have {
+		t.Errorf("want %+v, have %+v", want, have)
+	}
+
+	ctx = newrelicutil.WithTransaction(ctx, txn)
```
Comment (mkabischev) [inline newrelicutil_test.go]: Let's split into two separate tests.

**Review by mkabischev — CHANGES_REQUESTED** [review]

File: newrelicutil.go (hunk 3 above)
Comment (izaurio) [reply newrelicutil.go]: Like https://godoc.org/github.com/newrelic/go-agent#WrapHandle

File: newrelicutil.go (hunk 4 above)
Comment (izaurio) [reply newrelicutil.go]: No, see last line in the block.

File: newrelicutil.go (hunk 3 above)
Comment (mkabischev) [reply newrelicutil.go]: Ok, let's use the same name `WrapHandle`.

File: newrelicutil_test.go (hunk 7 above)
Comment (izaurio) [reply newrelicutil_test.go]: In this case I can't check `txn` value in `ctx` because key name is private.

File: newrelicutil.go (hunk 1 above)
Comment (izaurio) [reply newrelicutil.go:4]: done

File: newrelicutil_test.go (hunk 5 above)
Comment (izaurio) [reply newrelicutil_test.go:4]: done

**Review by mkabischev — APPROVED** [review]

**Pull request merged**